core/src/message/init.rs

```diff
@@ -161,8 +157,12 @@ impl InitPacket {
     }
 
     /// Packet destination (newly created program id).
-    pub fn destination(&self) -> ProgramId {
-        self.program_id
+    pub fn destination(&self, message_id: MessageId, nonce: Option<usize>) -> ProgramId {
```

Comment:
1) you need to have predefined id
2) you need to declare different constructors for in-program and in-extrinsics generations

Reply:
I cannot have a predetermined id if it requires a messageId and nonce, this can only be called once those are already made, such as in the case of the InitMessage

Reply:
> I cannot have a predetermined id if it requires a messageId and nonce, this can only be called once those are already made, such as in the case of the InitMessage

You will have pre-defined id if you follow the logic described here https://github.com/gear-tech/gear/pull/3010#discussion_r1301975721. Not sure whether that will be possible for the pallet level. I guess 2 constructors serve that purpose.